pkg/planner/property/physical_property.go

```diff
@@ -297,6 +305,75 @@ func (p *PhysicalProperty) IsSubsetOf(keys []*MPPPartitionColumn) []int {
+//     In this case, we can see that the child supplied partition keys is subset of parent required partition cols.
+func (p *PhysicalProperty) NeedEnforceExchangerWithHashByEquivalence(keys []*MPPPartitionColumn) bool {
+	// keys is the HashCol. If the partition cols are a subset of the hash cols, then need to enforce exchange.
+	if len(p.MPPPartitionCols) < len(keys) {
+		return true
+	}
+	uniqueID2MppCol := make(map[*MPPPartitionColumn]intset.FastIntSet, len(keys))
+	// for each partition column, we calculate the equivalence alternative closure of it.
+	for _, pCol := range p.MPPPartitionCols {
+		uniqueID2MppCol[pCol] = p.FD.ClosureOfEquivalence(intset.NewFastIntSet(int(pCol.Col.UniqueID)))
+	}
+
+	// there is a subset theorem here, if the child supplied keys is a subset of parent required mpp partition cols,
+	// the mpp partition exchanger can also be eliminated.
+	isSubset := true
+SubsetLoop:
+	for _, key := range keys {
+		for pCol, equivSet := range uniqueID2MppCol {
+			if checkEquivalence(equivSet, key, pCol) {
+				// yes, child can supply the same col partition prop. continue to next child supplied key.
+				continue SubsetLoop
+			}
+		}
+		// once a child supplied keys can't find direct/in-direct equiv all parent required partition cols,
+		// we can break the subset check.
+		isSubset = false
+		break SubsetLoop
+	}
+	// it's subset case, we don't need to add exchanger.
+	if isSubset {
+		return false
+	}
+
+	// when arrived here, it means the child supplied partition cols is not subset of parent required partition cols.
+	// then we need to check if the child supplied partition cols can supply the every partition col as parent required.
+	// iterate the parent prop required partition cols, check if it is compatible with children supplied.
+NextRequiredCol:
+	for pCol, equivSet := range uniqueID2MppCol {
+		// iterate the child supplied keys.
+		for _, key := range keys {
+			if checkEquivalence(equivSet, key, pCol) {
+				// yes, child can supply the same col partition prop. continue out to next required partition col.
+				continue NextRequiredCol
+			}
+		}
+		// if all child supplied keys iterated can NOT supply the same col partition prop, then we need to enforce exchange.
+		return true
+	}
+	// all required keys are satisfied, enforcer is not needed.
+	return false
+}
+
+func checkEquivalence(equivSet intset.FastIntSet, key, pCol *MPPPartitionColumn) bool {
+	// if the equiv set contain the key, it means it can supply the same col partition prop directly or in-indirectly.
+	// according to the old logic, when the child can supply the same key partition, we should check its collate-id
+	// when the new collation is enabled suggested by the collateID is negative. Or new collation is not set.
+	return equivSet.Has(int(key.Col.UniqueID)) &&
+		((key.CollateID < 0 && pCol.CollateID == key.CollateID) || key.CollateID >= 0)
```

Comment:
I check the code of collateId. The normal collatedId is positive. So why we check this collateId equivalence when key.collateid<0 ?

Reply:
<img width="1057" alt="image" src="https://github.com/user-attachments/assets/fb8044cb-142d-4c37-a86c-dfe09e7a57c5" />

pkg/util/collate/collate.go:120

Historically, TiDB had two encoding modes. In the first generation, all collation IDs were positive. However, once switched, they would be forcibly converted to negative values.

Reply:
In the first generation of collation in TiDB, there was direct syntactic support for collation. However, there was no actual implementation of collation itself, so no judgment was needed.